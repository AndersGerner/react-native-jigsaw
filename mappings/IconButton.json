{"name":"Icon Button","tag":"IconButton","category":"deprecated","preview_image_url":"https://res.cloudinary.com/altos/image/upload/draftbit/library/jigsaw-1.0/reps/Button_Icon.png","props":{"icon":{"label":"Icon Name","description":"Name of icon","editable":true,"required":false,"type":"icon","value":null},"size":{"label":"Icon Size","description":"Size of icon","editable":true,"required":false,"type":"flatArray","value":32,"options":[16,24,32]},"color":{"label":"Color","description":"Color of the icon","type":"color","value":"strong","editable":true,"required":true},"onPress":{"label":"Action","description":"Action to execute when icon button pressed","editable":true,"type":"action","value":null}},"layout":{"width":32,"height":32}}