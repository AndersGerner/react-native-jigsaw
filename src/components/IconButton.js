/* @flow */

import * as React from "react"
import { View, StyleSheet, ActivityIndicator } from "react-native"
import color from "color"

import Touchable from "./Touchable"
import Icon from "./Icon"
import { withTheme } from "../core/theming"
import type { IconSource } from "./Icon"
import type { Theme, $RemoveChildren } from "../types"
import { FORM_TYPES, COMPONENT_TYPES } from "../core/component-types"

type Props = $RemoveChildren<typeof Touchable> & {|
  /**
   * Icon to display.
   */
  icon: IconSource,
  /**
   * Color of the icon.
   */
  color?: string,
  /**
   * Size of the icon.
   */
  size?: number,
  /**
   * Whether the button is disabled. A disabled button is greyed out and `onPress` is not called on touch.
   */
  disabled?: boolean,
  /**
   * Whether the button is loading. A loading button is greyed out and `onPress` is not called on touch.
   */
  loading?: boolean,
  /**
   * Accessibility label for the button. This is read by the screen reader when the user taps the button.
   */
  accessibilityLabel?: string,
  /**
   * Function to execute on press.
   */
  onPress?: () => mixed,
  style?: any,
  /**
   * @optional
   */
  theme: Theme
|}

/**
 * An icon button is a button which displays only an icon without a label.
 *
 * <div class="screenshots">
 *   <figure>
 *     <img src="screenshots/icon-button-1.png" />
 *     <figcaption>Icon button</figcaption>
 *   </figure>
 *   <figure>
 *     <img src="screenshots/icon-button-2.png" />
 *     <figcaption>Pressed icon button</figcaption>
 *   </figure>
 * </div>
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { IconButton, Colors } from 'react-native-paper';
 *
 * const MyComponent = () => (
 *   <IconButton
 *     icon="add-a-photo"
 *     color={Colors.red500}
 *     size={20}
 *     onPress={() => console.log('Pressed')}
 *   />
 * );
 *
 * export default MyComponent;
 * ```
 */
const IconButton = ({
  icon,
  color: customColor,
  size = 32,
  accessibilityLabel,
  disabled,
  loading,
  onPress,
  theme,
  style,
  ...rest
}: Props) => {
  const iconColor = customColor || theme.colors.text
  const rippleColor = color(iconColor)
    .alpha(theme.disabledOpacity)
    .rgb()
    .string()

  return (
    <Touchable
      onPress={onPress}
      disabled={disabled || loading}
      background={Touchable.Ripple(rippleColor)}
      style={[
        styles.container,
        loading ||
          (disabled && {
            opacity: theme.disabledOpacity
          }),
        style
      ]}
      accessibilityLabel={accessibilityLabel}
      accessibilityTraits={disabled ? ["button", "disabled"] : "button"}
      accessibilityComponentType="button"
      accessibilityRole="button"
      accessibilityStates={disabled ? ["disabled"] : undefined}
      hitSlop={{ top: 6, left: 6, bottom: 6, right: 6 }}
      {...rest}>
      <View>
        {icon && loading !== true ? <Icon name={icon} size={size} color={iconColor} /> : null}
        {loading ? <ActivityIndicator size="small" color={iconColor} /> : null}
      </View>
    </Touchable>
  )
}

const styles = StyleSheet.create({
  container: {
    alignItems: "center",
    justifyContent: "center"
  }
})

export default withTheme(IconButton)

export const SEED_DATA = {
  name: "Icon Button",
  tag: "IconButton",
  category: COMPONENT_TYPES.deprecated,
  preview_image_url: "{CLOUDINARY_URL}/Button_Icon.png",
  props: {
    icon: {
      label: "Icon Name",
      description: "Name of icon",
      editable: true,
      required: false,
      type: FORM_TYPES.icon,
      value: null
    },
    size: {
      label: "Icon Size",
      description: "Size of icon",
      editable: true,
      required: false,
      type: FORM_TYPES.flatArray,
      value: 32,
      options: [16, 24, 32]
    },
    color: {
      label: "Color",
      description: "Color of the icon",
      type: FORM_TYPES.color,
      value: "strong",
      editable: true,
      required: true
    },
    onPress: {
      label: "Action",
      description: "Action to execute when icon button pressed",
      editable: true,
      type: FORM_TYPES.action,
      value: null
    }
  },
  layout: {
    width: 32,
    height: 32
  }
}
