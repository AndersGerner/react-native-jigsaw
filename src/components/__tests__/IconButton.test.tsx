import * as React from "react";
import renderer from "react-test-renderer";
import IconButton from "../IconButton";

it("renders icon button", () => {
  const tree = renderer.create(<IconButton onPress={() => {}} />).toJSON();

  expect(tree).toMatchSnapshot();
});
