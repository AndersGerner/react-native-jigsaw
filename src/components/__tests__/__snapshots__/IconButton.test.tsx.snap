// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders icon button 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  focusable={true}
  hitSlop={
    Object {
      "bottom": 6,
      "left": 6,
      "right": 6,
      "top": 6,
    }
  }
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "opacity": 1,
    }
  }
>
  <View>
    <View />
  </View>
</View>
`;
